Implement node state list

lukicdarkoo opened the pull request:
We want to add two supervisor functions:
- `wb_supervisor_node_save_state(node, state_name)`: saves node's state (like translation, rotation...) and all children of the node with name `state_name` (string)
- `wb_supervisor_node_load_state(node, state_name)` loads node's state (like translation, rotation...) and all children of the node with name `state_name` (string)

This will allow a partial restart of the world, allowing us to run multiple reinforcement learning environments in a single instance of Webots.

This PR also fixes #2759

**Documentation**
https://www.cyberbotics.com/doc/reference/supervisor?version=feature-state-list#wb_supervisor_node_save_state

https://www.cyberbotics.com/doc/reference/supervisor?version=feature-state-list#resetreload-matrix

**Tasks**
- [x] Add  `state_name` to `WbNode::reset` and `WbNode::save`
- [x] Add Supervisor C API
- Add support for other languages:
  - [x] C++
  - [x] Python
  - [x] Java
  - [x] Matlab
  - [x] ROS
- [x] Documentation
- [x] Check `mPreviousRotation = mSavedRotations[id];` in `WbSolid.cpp`
- [x] Tests
- [x] ROS tests
- [x] Controller.def + WbLanguage.cpp
- [x] Changelog    


**Samples**
<details>
<summary>Sample controller</summary>

```python
import time
from controller import Supervisor

robot = Supervisor()
timestep = int(robot.getBasicTimeStep())

left_motor = robot.getDevice('left wheel motor')
right_motor = robot.getDevice('right wheel motor')
left_motor.setPosition(float('inf'))
right_motor.setPosition(float('inf'))
left_motor.setVelocity(2)
right_motor.setVelocity(2)


target_node = robot.getFromDef('TARGET_NODE')
for _ in range(50): robot.step(timestep)
target_node.saveState('my_state')
time.sleep(0.5)

for _ in range(50): robot.step(timestep)
target_node.saveState('my_state2')
time.sleep(0.5)

for _ in range(50): robot.step(timestep)
target_node.loadState('my_state2')
robot.step(timestep)
time.sleep(0.5)

left_motor.setPosition(float('inf'))
right_motor.setPosition(float('inf'))
left_motor.setVelocity(2)
right_motor.setVelocity(2)

for _ in range(50): robot.step(timestep)
target_node.loadState('my_state')
robot.step(timestep)
time.sleep(0.5)

for _ in range(50): robot.step(timestep)
left_motor.setPosition(float('inf'))
right_motor.setPosition(float('inf'))
left_motor.setVelocity(2)
right_motor.setVelocity(2)

target_node.loadState('__init__')

while robot.step(timestep) != -1:
    pass
```

</details>


**Description**
|  | **Reload** | **Reset from UI** | **Reset from Supervisor** | **Load node's state from Supervisor** | **Reset physics** | **Reset node's physics** |
|-|-|-|-|-|-|-|
| **Resets simulation time** | Yes | Yes | Yes | No | No | No |
| **Removes nodes** | Yes | Yes | Yes | No | No | No |
| **Restarts controller** | Yes | Yes | No | No | No | No |
| **Stops sounds** | Yes | Yes | Yes | No | No | No |
| **Resets seeds** | Yes | Yes | Yes | No | No | No |
| **Resets physics** | Yes | Yes | Yes | Scoped | Yes | Scoped |
| **Resets physics plugin** | Yes | Yes | Yes | No | No | No |
| **Resets all fields** | Yes | No | No | No | No | No |
| **Add removed nodes** | Yes | No | No | No | No | No |
| **[Brake](../reference/brake.md)** | Resets | Releases | Releases | Releases | Nothing | Nothing |
| **[Charger](../reference/charger.md)** | Resets | Resets the `battery` field and the `emissiveColor` field of the [Material](../reference/material.md) node of the first [Shape](../reference/shape.md) child node | Resets the `battery` field and the `emissiveColor` field of the [Material](../reference/material.md) node of the first [Shape](../reference/shape.md) child node | Loads the `battery` field and resets the emissiveColor field of the [Material](../reference/material.md) node of the first [Shape](../reference/shape.md) child node | Nothing | Nothing |
| **[Display](../reference/display.md)** | Resets | Clears | Clears | Clears | Nothing | Nothing |
| **[Emitter](../reference/emitter.md)/[Receiver](../reference/receiver.md)** | Resets | Clears the message queue | Clears the message queue | Clears the message queue | Nothing | Nothing |
| **[Joint](../reference/joint.md)/[Motor](../reference/motor.md)** | Resets | Resets the position, velocity, acceleration, available torque, and available force | Resets the position, velocity, acceleration, available torque, and available force | Loads the position, resets the velocity, acceleration, available torque, and available force | Nothing | Nothing |
| **[LED](../reference/led.md)** | Resets | If the first child is a Light node, it resets the `color` field and it switches the LED off. If the first child is a [Shape](../reference/shape.md) node, it resets the `emissiveColor` field of its [Material](../reference/material.md) node. | If the first child is a Light node, it resets the `color` field and it switches the LED off. If the first child is a [Shape](../reference/shape.md) node, it resets the `emissiveColor` field of its [Material](../reference/material.md) node. | If the first child is a Light node, it resets the `color` field and it switches the LED off. If the first child is a [Shape](../reference/shape.md) node, it resets the `emissiveColor` field of its [Material](../reference/material.md) node. | Nothing | Nothing |
| **[Lidar](../reference/lidar.md)** | Resets | Resets the position of the rotating head | Resets the position of the rotating head | Resets the position of the rotating head | Nothing | Nothing |
| **[Pen](../reference/pen.md)** | Resets | Cleans the painted textures | Cleans the painted textures | Cleans the painted textures | Nothing | Nothing |
| **[Propeller](../reference/propeller.md)** | Resets | Resets the slow helix and it's initial position | Resets the slow helix and it's initial position | Resets the slow helix and it's initial position | Nothing | Nothing |
| **[Robot](../reference/robot.md)** | Resets | Resets the `battery` field, removes all the supervisor labels, resets the nodes visibility, and restarts the controller | Resets the `battery` field, removes all the supervisor labels, and resets the nodes visibility | Loads the `battery` field, removes all the supervisor labels, and resets the nodes visibility | Nothing | Nothing |
| **[Solid](../reference/solid.md)** | Resets | Resets the `translation` and `rotation` fields, and the physics | Resets the `translation` and `rotation` fields, and the physics | Loads the `translation` and `rotation` fields, and resets the physics | Nothing | Nothing |
| **[Track](../reference/track.md)** | Resets | Resets the motor position and the `translation` field of the textureTransform node of the [Appearance](../reference/appearance.md) node of the first [Shape](../reference/shape.md) children node | Resets the motor position and the `translation` field of the textureTransform node of the [Appearance](../reference/appearance.md) node of the first [Shape](../reference/shape.md) children node | Loads the motor position and the `translation` field of the textureTransform node of the [Appearance](../reference/appearance.md) node of the first [Shape](../reference/shape.md) children node | Nothing | Nothing |
| **[Viewpoint](../reference/viewpoint.md)** | Resets | Resets `orientation`, `position`, `near`, `far`, and `fieldOfView` fields | Resets `orientation`, `position`, `near`, `far`, and `fieldOfView` fields | Loads `orientation`, `position`, `near`, `far`, and `fieldOfView` fields | Nothing | Nothing |

omichel commented:
*Note:* Instead of naming the second argument `id` which makes me think of an integer value, I would rather name it `state_name` which make me think of a string.

lukicdarkoo commented:
@cyberbotics/maintainers It is ready for the first review (like design choices and obvious mistakes).

omichel commented on docs/reference/supervisor.md:
```diff
@@ -1284,6 +1284,8 @@ The last three are respectively the angular velocities around the x, y and z axe
 ---
 
+#### `wb_supervisor_node_save_state`
+#### `wb_supervisor_node_reset_state`
```
I would call this function `wb_supervisor_node_load_state` so that we better understand it is the counterpart of `_save_state`.

omichel commented on docs/reference/supervisor.md:
```diff
@@ -1358,7 +1372,16 @@ wb_supervisor_node_reset_physics(node)
+The `wb_supervisor_node_reset_state` function loads the state of the given node and all its descendant nodes.
+The function parameter `state_name` has to be a valid value, or the controller will fail.
+The valid `state_name` value means that the state is saved with the `wb_supervisor_node_reset_state` function, or a special value `__init__`.
+If the `node` parameter is equal to the root node and the `state_name` parameter is equal to `__init__` then the function will have a similar effect to the `wb_supervisor_simulation_reset` function, except it will not affect the user interface (for example, it will not reset the timer).
```
We should explain what is a state. Does it include translation, rotation, inertia, appearance, etc?

omichel commented on src/webots/nodes/WbSolid.hpp:
```diff
@@ -462,8 +462,8 @@ protected slots:
   WbRotation mPreviousRotation;  // used only by WbDifferentialWheels
-  WbVector3 mTranslationLoadedFromFile;
-  WbRotation mRotationLoadedFromFile;
+  QMap<QString, WbVector3> mTranslationLoadedFromFile;
+  QMap<QString, WbRotation> mRotationLoadedFromFile;
```
These member variables should probably be renamed to match their new purpose.

omichel commented on src/webots/nodes/WbTrack.hpp:
```diff
@@ -92,7 +92,7 @@ protected slots:
   WbShape *mShape;
   WbTextureTransform *mTextureTransform;
-  WbVector2 mInitialTextureTransformTranslation;
+  QMap<QString, WbVector2> mInitialTextureTransformTranslations;
```
A renaming of this member variable should be considered here as well (and maybe other member variables at other places).

omichel commented on src/webots/nodes/WbViewpoint.hpp:
```diff
@@ -170,14 +170,14 @@ class WbViewpoint : public WbBaseNode {
-  double mInitialFar;
-  double mInitialOrthographicHeight;
-  QString mInitialFollow;
+  QMap<QString, double> mInitialFieldOfView;
+  QMap<QString, WbVector3> mInitialPosition;
+  QMap<QString, WbRotation> mInitialOrientation;
+  QMap<QString, QString> mInitialDescription;
+  QMap<QString, double> mInitialNear;
+  QMap<QString, double> mInitialFar;
+  QMap<QString, double> mInitialOrthographicHeight;
+  QMap<QString, QString> mInitialFollow;
```
`mInitial` &rarr; `mSaved` for example.

omichel reviewed (commented):
It looks pretty good to me. I like the idea of re-using the existing reset code and extending it with QMaps.

lukicdarkoo commented:
ROS tests are passing now. I will backport the fix to the master/released branches as well.

omichel commented:
I am afraid that `wb_supervisor_node_load_state()` doesn't do what a user would expect: it will restore the pose (translation, rotation, joint position) of the node saved with `wb_supervisor_node_save_state()` and simply reset the node instead of restoring the supposedly saved state (e.g., color of a LED, image of a Display, content of an Emitter buffer, etc.). Implementing a correct `wb_supervisor_node_save/load_state()` functionality is much more difficult than what was done in this PR and maybe not what we want to do. Instead I would suggest to:
- rename these functions to `wb_supervisor_node_save_pose()` and `wb_supervisor_node_load_pose()`.
- change their behavior so that they will only save/load the pose of the nodes, e.g., translations, rotations and joint positions.
- document them properly.

lukicdarkoo commented:
For our use-case, resetting LED, the image of a Display, the content of an Emitter, and similar is not important. If the state of the corresponding controller Is not properly handled it will have a hard time handle all mentioned saved states.

Since we want to save/load more than pose I suggest that we keep the `wb_supervisor_node_save/load_state()` name, but to properly document the behavior. Here is the analysis of the current behavior.

The following states are properly saved/loaded:
| feature | suitable for RL |
| --- | --- |
| solid pose | suitable |
| joint positions | suitable |
| battery status | suitable |
| track position | not important  |
| connector status  | suitable |
| viewpoint (near, far, fieldOfView, position, orientation, follow) | not important |

The following states are reset:
| feature | suitable for RL |
| --- | --- |
| solid physics | suitable |
| led | not important |
| brake | not important |

In comparison to the `wb_supervisor_simulation_reset` function the following behavivors differ:
| feature | suitable for RL |
| --- | --- |
| no fields are reset |  should be changed --> TODO |
| timer is not reset | suitable |
| physics plugin is not reset | suitable |
| removed nodes are not re-added | should be changed --> TODO |
| seed is not updated | maybe suitable |

Let me know if you agree with the proposed changes (TODOs).

stefaniapedrazzi commented:
Just note that the advantage of "reset" vs "reload" is that it is much faster.
We should make sure that resetting fields and re-adding nodes won't slow down the simulation reset.
Otherwise this should be an option.

omichel commented:
That seems good to me and makes me think that the implementation of `wb_supervisor_node_reset()` could be as simple as this:
```C
#define wb_supervisor_node_reset(node) wb_supervisor_node_load_state(node, "__init__")
```

lukicdarkoo commented:
> Just note that the advantage of "reset" vs "reload" is that it is much faster.
> We should make sure that resetting fields and re-adding nodes won't slow down the simulation reset.
> Otherwise this should be an option.

I have just updated the table, added a new one for the world reloading, it should be clearer now (to me as well).

At the moment reset has a limited scope, doesn't reset fields, and doesn't re-add nodes. As the save/load state is somehow a subset of the reset maybe we should not include the fields reset and nodes re-adding.

omichel commented:
> At the moment reset has a limited scope, doesn't reset fields, and doesn't re-add nodes. As the save/load state is somehow a subset of the reset maybe we should not include the fields reset and nodes re-adding.

Yes, that would make this PR almost ready. And we could implement it later if needed. Meanwhile it can be achieved by supervisor functions as well. But whatever we decide, the documentation should be extremely well done to explain in details what the functions do and what they don't do.

omichel commented on docs/reference/supervisor.md:
```diff
@@ -2198,6 +2222,41 @@ The `wb_supervisor_simulation_reset` function sends a request to the simulator p
+| **Resets physics plugin** | Yes | Yes | Yes | No | No | No |
+| **Resets all fields** | Yes | No | No | No | No | No |
+| **Adds removed nodes** | Yes | No | No | No | No | No |
+| **[Brake](brake.md)** | Resets | Releases | Releases | Releases | Nothing | Nothing |
```
Instead of "Nothing" I would use "N/A" or "-" as "Nothing" let me think that it doesn't affect the node at all, but in reality it affects the physics of the node (but we already said it before). Ideally we should said "It doesn't change anything but the physics for this node", so maybe "N/A" (not applicable) or simply "-" means that we don't apply anything special in case of this node.

lukicdarkoo replied:
`N/A` is good

omichel reviewed (changes requested):
I like very much the idea of this table. I makes things way clearer. Also, it is good to have in the reference manual rather that in the guide. I have a few questions and suggestions though. Also, I am wondering it we shouldn't [implement it in HTML](https://stackoverflow.com/questions/46621765/can-i-merge-table-rows-in-markdown) in order to be able to merge some columns, that would certainly make it more readable and shorter.

lukicdarkoo commented on docs/reference/supervisor.md:
```diff
@@ -2198,6 +2222,41 @@ The `wb_supervisor_simulation_reset` function sends a request to the simulator p
+| - | **Reload** | **Reset from user interface** | **Reset from [Supervisor](supervisor.md)** | **Load node's state from Supervisor** | **Reset physics** | **Reset node's physics** |
+|-|-|-|-|-|-|-|
+| **Resets simulation time** | Yes | Yes | Yes | No | No | No |
+| **Removes nodes** | Yes | Yes | Yes | No | No | No |
```
It is correct, try to:
- run the simulation,
- add a node, and
- reset the simulation.

The added node will be removed.

omichel replied:
All right. I was wrong. Thanks for checking it.

lukicdarkoo commented:
> I am wondering it we shouldn't implement it in HTML in order to be able to merge some columns

This is a good idea. A long text repeats in a few rows.

omichel reviewed (approved):
Very nice, thank you.
